File: src/main/scala/eu/unitn/disi/distance/metrics/GreedyAlg.scala
package eu.unitn.disi.distance.metrics

import eu.unitn.disi.distance.utilities.DistanceWrapper
import org.apache.spark.sql.{Dataset, SparkSession}
import org.apache.spark.sql.functions.{asc, col}

import scala.collection.mutable
import scala.collection.mutable.ListBuffer

class GreedyAlg(spark: SparkSession,
                distances: Dataset[DistanceWrapper],
                dim: Int)
    extends IAlg(spark: SparkSession,
                 distances: Dataset[DistanceWrapper],
                 dim: Int) {

  override def computeMatching(): Array[DistanceWrapper] = {
    val start = System.currentTimeMillis()
    val matching: Array[Int] = Array.fill[Int](dim)(-1)
    val used: Array[Boolean] = Array.fill[Boolean](dim)(false)

    val zeros = distances.filter(col("distance") === 0.0).collect()
//    val broadcast = spark.sparkContext.broadcast(zeros)
    for (elem <- zeros) {
      matching(elem.idA) = elem.idB
      used(elem.idB) = true
      matches = matches ++ Seq(elem)
    }

    var checkingEmpty = distances
      .filter(col("distance") =!= 0.0)

    checkingEmpty = if (zeros.length <= 0) {
      checkingEmpty
    } else {
      checkingEmpty.filter(
        !col("idA")
          .isin(zeros.map(_.idA): _*)
          .or(!col("idB").isin(zeros.map(_.idB): _*)))
    }

    checkingEmpty.cache()

    val distancesList: Array[DistanceWrapper] =
      checkingEmpty.sort(asc("distance")).repartition(3000).collect()

    timeToCollect = System.currentTimeMillis() - start

    log.info(s"Collected distances in ${timeToCollect}ms")

    for (elem <- distancesList if matching(elem.idA) == -1 && !used(elem.idB)) {
      matching(elem.idA) = elem.idB
      used(elem.idB) = true
      matches = matches ++ Seq(elem)
    }

    var missingFst = ListBuffer.empty[Int]
    var missingSnd = mutable.Queue.empty[Int]
    for (i <- matching.indices) {
      if (matching(i) == -1) {
        missingFst += i
      }
      if (!used(i)) {
        missingSnd += i
      }
    }

//    println(s"MISSING FST ${missingFst.length}")
//    println(s"MISSING SND ${missingSnd.length}")
//
//    val distancesMap: Map[(Int, Int), DistanceWrapper] = fromListToMap(
//      distancesList)
//    for (elemFst <- missingFst) {
//      val elemSnd = missingSnd.dequeue()
//      matching(elemFst) = elemSnd
//      used(elemSnd) = true
//      println(
//        distancesMap.getOrElse((elemFst, elemSnd),
//                               DistanceWrapper(elemFst, elemSnd, 1.0)))
//      matches = matches ++ Seq(
//        distancesMap.getOrElse((elemFst, elemSnd),
//                               DistanceWrapper(elemFst, elemSnd, 1.0)))
//    }
    matches
  }
}

File: dm/algorithms/GreedyAlg.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import numpy as np
from pyspark.sql import functions as F
from .AbstractDistanceAlg import AbstractDistanceAlg


class AuctionAlg(AbstractDistanceAlg):

    def __init__(self, df, size):
        super().__init__(df, size)

    def compute_matching(self):
        assignments = np.full(self.size, -1)
        used = np.zeros(self.size, dtype=bool)

        zeros = self.df.filter(F.col('distance') == 0.0).collect()
        for row, col, dist in zeros:
            assignments[row] = col
            used[col] = True
            self.matches.append((row, col, dist))

        empty = self.df.filter(F.col('distance') != 0.0)

        if len(zeros) > 0:
            fst_ids = [elem[0] for elem in zeros]
            snd_ids = [elem[1] for elem in zeros]
            empty = empty.filter(not F.col('idA').isin(fst_ids)).filter(not F.col('idB').isin(snd_ids))

        empty.cache()

        distances = empty.sort(F.asc('distance')).repartition(3000).collect()

        for row, col, dist in distances:
            if assignments[row] == -1 and not used[col]:
                assignments[row] = col
                used[col] = True
                self.matches.append((row, col, dist))

        return self.matches
